datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -90,18 +90,12 @@ fn array_array(args: &[&dyn Array]) -> Result<ArrayRef> {
 
 /// put values in an array.
 pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
-    let arrays: Vec<&dyn Array> = values
-        .iter()
-        .map(|value| {
-            if let ColumnarValue::Array(value) = value {
-                Ok(value.as_ref())
-            } else {
-                Err(DataFusionError::NotImplemented(
-                    "Array is not implemented for scalar values.".to_string(),
-                ))
-            }
-        })
-        .collect::<Result<_>>()?;
-
-    Ok(ColumnarValue::Array(array_array(&arrays)?))
+    let mut arrays: Vec<ArrayRef> = vec![];
+    for x in values {
+        match x {
+            ColumnarValue::Array(array) => arrays.push(array.clone()),
+            ColumnarValue::Scalar(scalar) => arrays.push(scalar.to_array().clone()),
+        }
+    }
```

Comment:
If you wanted to write this slightly more idomatically it could be

```suggestion
    let arrays: Vec<ArrayRef> = values.iter().map(|x| {
        match x {
            ColumnarValue::Array(array) => array.clone(),
            ColumnarValue::Scalar(scalar) => scalar.to_array().clone(),
        })
        .collect()
    }
```

However, I don't think it makes any practical difference